test/Fencer/Rules/Test.hs

```diff
@@ -71,6 +81,87 @@ test_loadRulesRecursively =
         (sortOn domainDefinitionId [domain1, domain2])
         (sortOn domainDefinitionId definitions)
 
+-- | Test that a rule limit unit change adds a new counter and leaves
+-- the old one intact.
+--
+-- TODO(md)-2019-10-11: Sometimes this test non-deterministically fails with:
+--
+-- Got wrong gRPC error response
+--         expected: ClientIOError (GRPCIOBadStatusCode StatusUnknown
+--           (StatusDetails {unStatusDetails = "rate limit descriptor
+--             list must not be empty"}))
+--          but got: ClientIOError (GRPCIOBadStatusCode StatusUnavailable
+--            (StatusDetails {unStatusDetails = "Endpoint read failed"}))
+test_ruleLimitUnitChange :: TestTree
+test_ruleLimitUnitChange =
+  -- TODO(md): creating a server here sometimes clashes with executing
+  -- server tests concurrently, which occasionally leads to test
+  -- failures. Fix this either by making sure the port hasn't been
+  -- binded or in some other way.
+  withResource createServerAppState destroyServerAppState $ \ioLogIdState ->
+    testCase "A rule limit unit change on rule reloading" $ do
+      Temp.withSystemTempDirectory "fencer-config-unit" $ \tempDir -> do
+        createDirectoryIfMissing True (tempDir </> dir)
+
+        definitions1 <- writeLoad tempDir merchantLimitsText1
+        (_, _, state) <- ioLogIdState
+
+        atomically $ setRules state (mapRuleDefs definitions1)
+
+        ruleTree :: RuleTree <- atomically $ fromJust <$> StmMap.lookup domainId (appStateRules state)
```

Comment:
Oh, never use `fromJust`, `head`, `(!!)` and other partial functions from `base`. Even in tests. I'd say it's not a big deal to use partial functions, but those ones from `base` throw the call stack away and you get a useless error message and have to grep through the entire codebase and check each single invocation of the failed function.

Reply:
Does your appeal apply to the `base-prelude` base as well?